lib/Conversion/ExportVerilog/PrepareForEmission.cpp

```diff
@@ -46,6 +46,18 @@ bool ExportVerilog::isSimpleReadOrPort(Value v) {
   return isa<WireOp, RegOp>(readSrc);
 }
 
+// Check if the value is deemed worth spilling into a wire.
+static bool shouldSpillWire(Operation &op, const LoweringOptions &options) {
+  auto isAssign = [](Operation *op) { return isa<AssignOp>(op); };
+
+  // If there are more than the maximum number of terms in this single result
+  // expression, and it hasn't already been spilled, this should spill.
+  return isVerilogExpression(&op) &&
+         op.getNumOperands() > options.maximumNumberOfTermsPerExpression &&
+         op.getNumResults() == 1 &&
+         llvm::none_of(op.getResult(0).getUsers(), isAssign);
```

Comment:
And regardless if one use of it is an assignment, if it has a lot of terms, doesn't it need to be spilled?  assignments with huge rhs should trigger parser errors in tools.
In that spirit, isn't this "mustSpillWire"?  token limits are a correctness constraint on the output.

Reply:
This has to do with how prepare is treating the block as a worklist. The first time shouldSpillWire is called, the large expression may or may not be assigned to a wire. If it's not, we create the wire and assign it. Then in the next iteration through the worklist, the pointer is still pointing to the expression, but now it is spilled, so we move on.

This was more complicated and subtle than I initially expected... maybe there is a smarter way to do this, but this works.

> And regardless if one use of it is an assignment, if it has a lot of terms, doesn't it need to be spilled? assignments with huge rhs should trigger parser errors in tools.
In that spirit, isn't this "mustSpillWire"? token limits are a correctness constraint on the output.

There is a phase-ordering question related to splitting and spilling. What this PR is trying to do is spill under some conditions during prepare, which was happening before in export. If there is a huge expression that's written to a wire, it may also need to be split _and_ spilled into many wires... but this currently doesn't do that.